src/main/java/dev/sigstore/KeylessVerifier.java

```diff
@@ -0,0 +1,179 @@
+import java.sql.Date;
+import java.time.Instant;
+import java.util.Base64;
+import java.util.Optional;
+
+/** Verify hashrekords from rekor signed using the keyless signing flow with fulcio certificates. */
+public class KeylessVerifier {
+  private final FulcioVerifier fulcioVerifier;
+  private final RekorVerifier rekorVerifier;
+  private final RekorClient rekorClient;
+
+  private KeylessVerifier(
+      FulcioVerifier fulcioVerifier, RekorClient rekorClient, RekorVerifier rekorVerifier) {
+    this.fulcioVerifier = fulcioVerifier;
+    this.rekorClient = rekorClient;
+    this.rekorVerifier = rekorVerifier;
+  }
+
+  public static KeylessVerifier.Builder builder() {
+    return new KeylessVerifier.Builder();
+  }
+
+  public static class Builder {
+    private FulcioVerifier fulcioVerifier;
+    private RekorClient rekorClient;
+    private RekorVerifier rekorVerifier;
+
+    public KeylessVerifier.Builder fulcioVerifier(FulcioVerifier fulcioVerifier) {
+      this.fulcioVerifier = fulcioVerifier;
+      return this;
+    }
+
+    public KeylessVerifier.Builder rekorClient(
+        RekorClient rekorClient, RekorVerifier rekorVerifier) {
+      this.rekorClient = rekorClient;
+      this.rekorVerifier = rekorVerifier;
+      return this;
+    }
+
+    public KeylessVerifier build() {
+      Preconditions.checkNotNull(fulcioVerifier);
+      Preconditions.checkNotNull(rekorVerifier);
+      return new KeylessVerifier(fulcioVerifier, rekorClient, rekorVerifier);
+    }
+
+    public Builder sigstorePublicDefaults()
+        throws IOException, InvalidAlgorithmParameterException, CertificateException,
+            InvalidKeySpecException, NoSuchAlgorithmException {
+      var fulcioCert =
+          Resources.toByteArray(
+              Resources.getResource("dev/sigstore/tuf/production/fulcio_v1.crt.pem"));
+      var ctfePublicKey =
+          Resources.toByteArray(Resources.getResource("dev/sigstore/tuf/production/ctfe.pub"));
+      var rekorPublicKey =
+          Resources.toByteArray(Resources.getResource("dev/sigstore/tuf/production/rekor.pub"));
+      fulcioVerifier(FulcioVerifier.newFulcioVerifier(fulcioCert, ctfePublicKey));
+      rekorClient(RekorClient.builder().build(), RekorVerifier.newRekorVerifier(rekorPublicKey));
+      return this;
+    }
+  }
+
+  public KeylessVerificationResult verifyOnline(
+      byte[] artifactDigest, byte[] certChain, byte[] signature)
+      throws IOException, NoSuchAlgorithmException, InvalidKeyException, CertificateException,
+          SignatureException {
+    var signingCert = SigningCertificate.from(Certificates.fromPemChain(certChain));
+    var leafCert = signingCert.getLeafCertificate();
+
+    var checks = KeylessVerificationResult.builder();
+
+    try {
+      fulcioVerifier.verifyCertChain(signingCert);
+      checks.checkPassed(Check.FULCIO_CERT);
+    } catch (FulcioVerificationException ex) {
+      return checks.checkFailed(Check.FULCIO_CERT, ex.getMessage());
```

Comment:
if we're not collecting errors I'm a little confused why we bother with this checks collection at all.

Reply:
I mean this is what I was not feeling great about. Is there any point in collecting these things?